src/sqlfluff/core/rules/crawlers.py

```diff
@@ -0,0 +1,125 @@
+"""Definitions of crawlers."""
+
+from typing import Iterator, Set, cast
+from sqlfluff.core.parser.segments.base import BaseSegment
+from sqlfluff.core.parser.segments.raw import RawSegment
+
+from sqlfluff.core.rules.context import RuleContext
+
+
+class BaseCrawler:
+    """The base interface for crawler classes."""
+
+    def __init__(self, works_on_unparsable: bool = False, **kwargs):
+        self.works_on_unparsable = works_on_unparsable
+
+    def passes_filter(self, segment: BaseSegment):
+        """Returns true if this segment considered at all."""
+        return self.works_on_unparsable or not segment.is_type("unparsable")
+
+    def crawl(self, context: RuleContext) -> Iterator[RuleContext]:
+        """Yields a RuleContext for each segment the rule should process."""
+        raise NotImplementedError(  # pragma: no cover
+            "Use a sepecific crawler, not the BaseCrawler. If you're seeing "
+            "this error, a rule has likely not set it's crawl_behaviour attribute."
+        )
+
+
+class RootOnlyCrawler(BaseCrawler):
+    """A crawler that doesn't crawl.
+
+    This just yields one context on the parent segment of the file.
+    """
+
+    def crawl(self, context: RuleContext) -> Iterator[RuleContext]:
+        """Yields a RuleContext for each segment the rule should process."""
+        if not self.passes_filter(context.segment):
+            return  # pragma: no cover
+        yield context
```

Comment:
Simpler & avoids `no cover`:
```suggestion
        if self.passes_filter(context.segment):
            yield context
```